stake-pool/program/src/processor.rs

```diff
@@ -430,79 +673,377 @@ impl Processor {
+        for validator_stake_record in &mut validator_stake_list.validators {
+            if validator_stake_record.last_update_epoch >= clock.epoch {
+                continue;
+            }
+            for (validator_stake_account, validator_account) in validator_stake_accounts
+                .iter()
+                .zip(validator_accounts.iter())
+            {
+                if validator_stake_record.validator_account
+                    != validator_account.ok_or(StakePoolError::WrongStakeState)?
+                {
+                    continue;
+                }
+                validator_stake_record.last_update_epoch = clock.epoch;
+                validator_stake_record.balance = **validator_stake_account.lamports.borrow();
+                changes = true;
+            }
+        }
+
+        if changes {
+            validator_stake_list.serialize(&mut validator_stake_list_info.data.borrow_mut())?;
+        }
+
+        Ok(())
+    }
+
+    /// Processes `UpdatePoolBalance` instruction.
+    pub fn process_update_pool_balance(
+        _program_id: &Pubkey,
+        accounts: &[AccountInfo],
+    ) -> ProgramResult {
+        let account_info_iter = &mut accounts.iter();
+        // Stake pool account
+        let stake_pool_info = next_account_info(account_info_iter)?;
+        // Account storing validator stake list
+        let validator_stake_list_info = next_account_info(account_info_iter)?;
+        // Clock sysvar account
+        let clock_info = next_account_info(account_info_iter)?;
+        let clock = &Clock::from_account_info(clock_info)?;
+
+        // Get stake pool stake (and check if it is iniaialized)
         let mut stake_pool = State::deserialize(&stake_pool_info.data.borrow())?.stake_pool()?;
 
-        if *withdraw_info.key
-            != Self::authority_id(
-                program_id,
-                stake_pool_info.key,
-                Self::AUTHORITY_WITHDRAW,
-                stake_pool.withdraw_bump_seed,
-            )?
-        {
-            return Err(Error::InvalidProgramAddress.into());
+        // Check validator stake account list storage
+        if *validator_stake_list_info.key != stake_pool.validator_stake_list {
+            return Err(StakePoolError::InvalidValidatorStakeList.into());
+        }
+
+        // Read validator stake list account and check if it is valid
+        let validator_stake_list =
+            ValidatorStakeList::deserialize(&validator_stake_list_info.data.borrow())?;
+        if !validator_stake_list.is_initialized {
+            return Err(StakePoolError::InvalidState.into());
+        }
+
+        let mut total_balance: u64 = 0;
+        for validator_stake_record in validator_stake_list.validators {
+            if validator_stake_record.last_update_epoch < clock.epoch {
+                return Err(StakePoolError::UpdateStakeList.into());
+            }
+            total_balance += validator_stake_record.balance;
```

Comment:
I'm not convinced :)
The rent is a passive amount which does not get any rewards (afaik), but subtracting every time (and we need to do it every time, when validator stake account is added, when deposit is done, etc) will only make math more complicated. 

Unless of course we should be able to withdraw everything from the pool. In this case yes, we will need to calculate balance - rent. So when admin adds validator stake account with 1 SOL and 0.1 SOL is rent he should receive 0.9 POOL TOKENS (considering 1:1 math) in return. Because when withdrawing maximum of 0.9 SOL can be split from the stake account.

Is it what you mean?

Reply:
You make a really good point actually.

If someone decides to withdraw 100 times instead of just 1 time, they'll steal value from the pool in the form of the rent exemption, and it should be up to the withdrawer to pay for that rent exempt amount through their pool tokens.  In that case, pool tokens must be valued in terms of the rent exempt reserves.  Never mind then, use the full balance to calculate the value of the pool!